### ISSUE TITLE
Nightly Builds For Android GL

### ISSUE THREAD

**Issue by bleege** [mapbox/mapbox-gl-native#826]
Document current process for creating the Android GL SDK.  Figure out a way that it can be automated to produced nightly builds.

/cc @ljbade

**Comment by bleege** [comment]
#### Current Process

1. `make android` in `/`
  * Uses Mason and Android NDK to produce Shareable Object (.so) files for inclusion in Android GL SDK
  * Copies generated `.so` into `android/java/lib/src/jniLibs/-ARCH-/libmapbox-gl.so`
1. `./gradlew clean assembleRelease` in `android/java`
  * Produces `lib-release.aar` located in `android/java/lib/build/outputs/aar`

**Comment by bleege** [comment]
TODO: Research way to get Gradle to execute initial Makefile and then kickoff AAR generating / distribution afterwards.

**Comment by ljbade** [comment]
I cursory Google for "apklib jni" seems to indicate that apklib does support jni, but it is not clear if Maven plugin supports it. There is some discussion around whether or not to add it to Maven 3, mostly around how to handle JNI native dependencies. Not clear if it made it in or not.

**Comment by bleege** [comment]
Working on branch [nightlyBuilds](https://github.com/mapbox/mapbox-gl-native/compare/nightlyBuilds).

**Comment by bleege** [comment]
Successfully tested build process refactoring work!  The first SNAPSHOT release (aka what will become nightly builds) of Mapbox GL Android was pushed to Maven Central.  The process is creating the following artifacts:

* .aar Library File (primary library)
* .apklib (Legacy library for Eclipse users)
* Generated JavaDocs
* Source Code (Only Java, not GL Core)
* .pom File For Dependency Management

https://oss.sonatype.org/content/repositories/snapshots/com/mapbox/mapboxgl/mapbox-gl-android-sdk/0.0.1-SNAPSHOT/

**Comment by bleege** [comment]
Confirmed that the generated `.apklib` file contains the gl core .so files in `jniLibs`.  This should be all Eclipse users need to make including the SDK work as expected.

**Comment by ljbade** [comment]
@bleege Awesome, fingers crossed we don't hear from Eclipse users.

**Comment by bleege** [comment]
All the building blocks from a Gradle scripting perspective should be built now to automatically produce clean releases and publish them to Maven Central (both SNAPSHOTS and Releases).  Here's an overview of how I'm thinking things will flow:

1. All properties (Android targets, version numbers, etc) of the Mapbox GL Android SDK are configured via `android/java/MapboxGLAndroidSDK/gradle.properties`
  * Once configured will be committed to GitHub
1. Use existing Make file process to bundle up the JNI Shared Objects (On Travis CI)
  * [ ] Will need to create new Make Target `make android-deploy` that extends from `make android-all`
1. Make file will call `android/java/MapboxGLAndroidSDK/deploy.sh` at conclusion of Step 1 (On Travis CI)
  * Wrapper shell script for Gradle process that cleans, assembles, and publishes release
  * NOTE: Driven by value of `VERSION_NAME` in `android/java/MapboxGLAndroidSDK/gradle.properties`
    * `-SNAPSHOT` will produce snapshot builds
    * **Everything else will be processed as a production release**

Remaining Tasks:

* [ ] Put Mapbox Sonatype (aka Maven Central) credentials in a secured could environment that Travis can access
* [ ] Put Mapbox private / public RSA key (used for artifact signing) in a secured cloud environment that Travis can access
* [ ] Update `android/java/MapboxGLAndroidSDK/gradle.properties` to include secure access instructions for Sonatype credentials and RSA keys

These remaining tasks can be worked on after @willwhite gives his "AWS 101" presentation to the Mapbox Mobile team tomorrow.

In the meantime, this infrastructure can be merged into the `master` branch.

/cc @ljbade @incanus

**Comment by incanus** [comment]
:+1: 

> @willwhite gives his "AWS 101" presentation to the Mapbox Mobile team tomorrow.

`o_O`

>  a secured cloud environment that Travis can access

Cool, let me know if you need any help from me.

**Comment by ljbade** [comment]
@mikemorris I think @incanus commited the ios public key and the password protected private key into git? Do we want to do that too?

**Comment by willwhite** [comment]
It was great hearing about the details of Android builds. My takeaways/suggestions are:

- Use empty commit message with `[publish vx.x.x]` pattern to push releases. Kick off the build [if this pattern appears in the commit message](https://github.com/mapbox/mapbox-studio/blob/bc5eeac0d1e45595c203d13798d3d9fdccfee972/scripts/build-travis.sh#L7).
- Only do releases via travis - no human intervention.
- Store secrets like certificates with [`travis encrypt`](http://docs.travis-ci.com/user/encryption-keys/).
- Do as much as possible on travis and only boot an EC2 if you definitely need it. And if you need it, use a stock Ubuntu AMI whenever possible.

**Comment by bleege** [comment]
Removed `cleanJNIBuilds()` task from build.gradle as it was being erroneously included as a dependency for the `build()` task, causing #859.  Worked with @ljbade to debug and get working on Mac and Linux.

**Comment by bleege** [comment]
Thoughts About Android Code Signing And Distribution Security

1. Implement @willwhite's [suggestions](https://github.com/mapbox/mapbox-gl-native/issues/826#issuecomment-73997874)
  * Travis is only "person" allowed to sign Android resources as well as publish to Maven Central
  * Control Travis via Git commits
1. Encrypt and Store RSA Certificate in Git
1. Encrypt `gradle.properties` with passwords and path to RSA Certificate
  * Mirror [iOS process](https://github.com/mapbox/mapbox-gl-native/blob/master/.travis.yml#L75-L80)
  * [ ] Determine process for updating release numbers that's normally part of `gradle.properties`

**Comment by bleege** [comment]
Rules and options for configuring [Gradle properties].  Key take away:

> The configuration is applied in following order (if an option is configured in multiple locations the last one wins):
from gradle.properties in project build dir.
from gradle.properties in gradle user home.
from system properties, e.g. when -Dsome.property is set on the command line.

**Comment by incanus** [comment]
@bleege In the event that you're curious or it's otherwise useful to you, this was the guide I used for iOS signing and getting up to speed on Travis security: 

http://www.objc.io/issue-6/travis-ci.html

Again, feel free to ping me and I can show you, but just FYI.

**Comment by bleege** [comment]
Found a great overview for doing what we need to do: [Automatically Publish to Sonatype with Gradle and Travis CI](http://benlimmer.com/2014/01/04/automatically-publish-to-sonatype-with-gradle-and-travis-ci/).  The only potential pitfall is that they use Jenkins internally to actually do the publishing to Sonatype because the company isn't allowed to let keyless leave their internal network.  I think we'll be able to work around that though with Travis.

**Comment by bleege** [comment]
Updating Goals / Thinking On Publishing Strategy:

* Continuous Full Builds driven by commits to repo
* Nightly / On Demand SNAPSHOTS published to Maven Central SNAPSHOT platform
* Commit Driven Releases published to Maven Central (Staging) Release platform
  * NOTE: Humans need to walk the build from Staging To Release via [Sonotype Nexus Web UI](https://oss.sonatype.org)
* Keep `grade.properties` plan text / NOT encrypted for easier developer access
  * Done by using Travis based encrypted variables in command line options

**Comment by bleege** [comment]
Update:

After chatting with @ljbade I've discovered that while Travis kicks off Android builds for every commit, but the actual builds are done on EC2.  This means that the method described in [the blog link from last night ](http://benlimmer.com/2014/01/04/automatically-publish-to-sonatype-with-gradle-and-travis-ci/) won't work for us.

Next Steps:
Re-imagine how to get secret credentials to EC2 via Travis and GitHub without the rest of the world seeing.

**Comment by bleege** [comment]
Reviewed re-imagination solution with @willwhite this afternoon.  For the most part it looks good and should be secure.  He's going to check on one other possible option before I upload the last commit and continue down this path.

**Comment by willwhite** [comment]
Since we are already causing travis to boot up an EC2 to run out the build, lets keep our signing secrets on S3 and have the EC2 grab them before starting the build process. We take a similar approach in Mapbox Studio for [Windows signing keys](https://github.com/mapbox/mapbox-studio/blob/ba7025c610f1701acfe41380154c4fd9cadf3830/scripts/build-atom.sh#L105-L106), but there its happening on travis - not an EC2.

We can use [EC2 Roles](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html) to delegate temporary access to the EC2 so it can access the creds.

1. Augment the EC2 Role in [travis-resources.template](https://github.com/mapbox/mapbox-gl-native/blob/df57d6ee8b8eb04499385fa42466a8d165a9d829/travis/travis-resources.template#L122). [Docs for this are here](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-iam-role.html). The role should provide a policy that allows the EC2 to retrieve the signing files from an S3 location, like `s3://mapbox/mapbox-gl-native/android-key`.
2. Update the running CloudFormation stack (called `travis-mapbox-gl-native`) with the adjusted template [using the AWS CLI](http://aws.amazon.com/cli/).
3. When we launch the instance we already [assign the the role to the instance](https://github.com/mapbox/mapbox-gl-native/blob/master/android/scripts/run-build.sh#L54) via an instance profile.
4. Install [using the AWS CLI](http://aws.amazon.com/cli/) on the EC2 (if it's not there already) and using `aws s3 cp` to grab the remote files.

/cc @bleege @ljbade

**Comment by ljbade** [comment]
Should be easy

 @incanus You keep the iOS keys in git right? Do you think we should also put those on S3?

**Comment by ljbade** [comment]
Should be easy

 @incanus You keep the iOS keys in git right? Do you think we should also put those on S3?

**Comment by ljbade** [comment]
Should be easy

 @incanus You keep the iOS keys in git right? Do you think we should also put those on S3?

**Comment by bleege** [comment]
See #872 too.

**Comment by bleege** [comment]
This all sounds great @willwhite.  Thanks!

> Should be easy

I'm brand new to AWS / EC2 / S3.  Mind if I tag team on this with you today @ljbade?

**Comment by bleege** [comment]
See #875 too.

**Comment by bleege** [comment]
Just updated the Cloud Formation with the updated EC2 Role.  The install took...no errors from AWS.  Will no test the full stack process to see what happens / where it'll break.

**Comment by bleege** [comment]
Still waiting on Travis.... https://travis-ci.org/mapbox/mapbox-gl-native/builds/50697560

**Comment by bleege** [comment]
Confirmed that Android `debug` releases worked.  Exploring issues with `release` version now that Travis is unclogged.

**Comment by ljbade** [comment]
Yeah I'm gonna have to wait for your merge for the `move-buckets` branch cause I need to pull in your changes to the cloudformation template.

**Comment by bleege** [comment]
Putting issue on hold for a day or so as @kkaefer is taking a run at [integrating the Core C++ with Gradle / Android Studio](http://www.shaneenishry.com/blog/2014/08/17/ndk-with-android-studio/).  The primary reason for this effort is to enable *much better* C++ debugging process in GL Android SDK via Android Studio.

If this is successful, it'll also force us to rethink the Automated Build Process as it'll fully encapsulate the build process within Gradle.  This will allow full separation from the existing Makefile process.  In theory we'll still be able to build with Travis CI, but it also might make us want to look at [Jenkins CI](http://jenkins-ci.org) either on our own [AWS Cloud Formation powered EC2 instance](https://wiki.jenkins-ci.org/display/JENKINS/AWS+Cloudformation+Plugin) or a [hosting provider](https://www.cloudbees.com).

Resources:

"NDK With Android Studio"
http://www.shaneenishry.com/blog/2014/08/17/ndk-with-android-studio/

"Building Android NDK projects with Gradle and Travis CI"
http://www.armedpineapple.co.uk/2014/01/building-android-ndk-projects-with-gradle-and-travis-ci/

/cc @mapbox/gl @incanus @ljbade @willwhite @freenerd

**Comment by ljbade** [comment]
@bleege Where did we get to with this?

**Comment by ljbade** [comment]
@bleege Where did we get to with this?

**Comment by ljbade** [comment]
@bleege Where did we get to with this?

**Comment by ljbade** [comment]
@bleege Where did we get to with this?

**Comment by ljbade** [comment]
@bleege Where did we get to with this?

**Comment by ljbade** [comment]
@bleege Where did we get to with this?

**Comment by ljbade** [comment]
@bleege Where did we get to with this?

**Comment by ljbade** [comment]
@bleege Where did we get to with this?

**Comment by ljbade** [comment]
@bleege Where did we get to with this?

**Comment by ljbade** [comment]
@bleege Where did we get to with this?

**Comment by ljbade** [comment]
@bleege Where did we get to with this?

**Comment by ljbade** [comment]
@bleege Where did we get to with this?

**Comment by ljbade** [comment]
@bleege Where did we get to with this?

**Comment by ljbade** [comment]
@bleege Where did we get to with this?

**Comment by ljbade** [comment]
@bleege Where did we get to with this?

**Comment by ljbade** [comment]
@bleege Where did we get to with this?

**Comment by bleege** [comment]
@ljbade It's been on hold waiting feedback from @kkaefer on his thoughts regarding C++ and DJINNI setup from last week.  @bsudekum is also talking with the Travis people today / tomorrow regarding improved service.  Long story short, things looked like they were changing imminently so it didn't make sense to keep pushing on (especially with the clogged Travis pipes).

Unless it's a blocker to you today, my thought is to give it another day to hear what @kkaefer and @bsudekum have to say.  Sound good?

**Comment by ljbade** [comment]
Yeah sure, I was just going through Android tickets to work out what to work on next.

**Comment by ljbade** [comment]
Latest ticket https://github.com/mapbox/mapbox-gl-native/issues/1672

**Comment by mb12** [comment]
@ljbade Is it possible to publish a signed apk for sample app as part of nightly builds? Since Android allows OTA installs of apps, this would make it very easy to get feedback/testing on multiple devices.

**Comment by bleege** [comment]
@mb12 We're more likely to publish nightly artifacts (`.aar`, `.apklib`, etc) and not actual apps.

**Comment by kkaefer** [comment]
We could use [buck's API](http://buckbuild.com/concept/http_cache_api.html) for uploading build artifacts.
